### PR TITLE
Spa thalamus routing fix

### PR DESCRIPTION
**Pull request by tcstewar**
spa.Actions that caused routing to a module that has multiple inputs didn't work.  For example:

```python
spa.Actions('dot(buffer, A) --> cmp_A=A, cmp_B=B')
```

This was just due to ```spa.get_module``` not handling this case.  This PR fixes it and adds a test to make sure it works.

### CODE REVIEW

**Comment by xchoo** [comment]
Flake8's complaining (unused import). XD

**Comment by tcstewar** [comment]
Oops...  damnit, pep8 doesn't check for that and that's what I've been checking against...  oops..  let's see how this version goes.

File: nengo/spa/spa.py [hunk 1]
```diff
@@ -88,7 +88,11 @@ def __setattr__(self, key, value):
 
     def get_module(self, name):
         """Return the module for the given name."""
-        return self._modules[name]
+        if name in self._modules:
+            return self._modules[name]
+        elif '_' in name:
+            module, name = name.rsplit('_', 1)
```
Comment (xchoo) [inline nengo/spa/spa.py]: So this means that we can't have inputs and outputs names with an underscore in them, yes? This should be documented somewhere. (If we use an lsplit here, then we can't use module names with underscores.. haha)

File: nengo/spa/thalamus.py [hunk 2]
```diff
@@ -188,6 +189,8 @@ def add_route_effect(self,
             subdim = self.subdim_channel
             if isinstance(target_module, nengo.spa.Buffer):
                 subdim = target_module.state.dimensions_per_ensemble
+            elif isinstance(source_module, nengo.spa.Buffer):
+                subdim = source_module.state.dimensions_per_ensemble
```
Comment (xchoo) [inline nengo/spa/thalamus.py]: On a related note, what if it connects to ```spa.Memory```. Should there be logic for that here too?

File: nengo/spa/tests/test_thalamus.py [hunk 3]
```diff
@@ -127,6 +127,56 @@ def input_func(t):
...
+
+        node1 = nengo.Node([0, 1, 0])
+        node2 = nengo.Node([0, 0, 1])
+
+        nengo.Connection(node1, model.buff1.state.input)
+        nengo.Connection(node2, model.buff2.state.input)
+
+        actions = spa.Actions('dot(ctrl, A) --> cmp_A=buff1, cmp_B=buff1',
+                              'dot(ctrl, B) --> cmp_A=buff1, cmp_B=buff2',
+                              'dot(ctrl, C) --> cmp_A=buff2, cmp_B=buff2',
```
Comment (xchoo) [inline nengo/spa/tests/test_thalamus.py]: I'd also like to see a test where the source is non-default option.

File: nengo/spa/spa.py (hunk 1 above)
Comment (tcstewar) [inline nengo/spa/spa.py]: Yup, that should be documented somewhere.  In fact, I really don't like the underscore convention here, but nothing else was coming to mind.

File: nengo/spa/tests/test_thalamus.py (hunk 3 above)
Comment (tcstewar) [inline nengo/spa/tests/test_thalamus.py]: > I'd also like to see a test where the source is non-default option.

I don't think we have any built-in modules with multiple sources....

File: nengo/spa/thalamus.py (hunk 2 above)
Comment (tcstewar) [inline nengo/spa/thalamus.py]: ```spa.Memory``` is a subclass of ```spa.Buffer```.  So no logic needed  :)

File: nengo/spa/tests/test_thalamus.py (hunk 3 above)
Comment (xchoo) [inline nengo/spa/tests/test_thalamus.py]: I think the '.' operator will work, but I'm not sure how well it will play with the way we are evaluating the action strings.

Comment (tcstewar) [inline nengo/spa/tests/test_thalamus.py]: That does feel like a better option, and it might just be a simple option to add to that evaluation system.  But I haven't looked at that code since @arvoelke rewrote it all.  :)

File: nengo/spa/spa.py (hunk 1 above)
Comment (jgosmann) [inline nengo/spa/spa.py]: Would `.` instead of `_` work? It should not occur as part of a module name, should it? And it seems like the natural syntax to me.

Comment (tcstewar) [inline nengo/spa/spa.py]: Yes, it would work, but it would require adding that functionality to the spa/action_objects system, and that's well outside the scope of this PR.

Note: this PR is about fixing a bug in the spa thalamus system that failed on ```module_whatever``` situations, which is what the rest of the spa system was built with.  It'd be a separate PR to go change everything over to ```module.whatever``` instead of ```module_whatever```.

File: nengo/spa/tests/test_thalamus.py (hunk 3 above)
Comment (xchoo) [inline nengo/spa/tests/test_thalamus.py]: Whoops. Added the line note to the wrong line! Haha
Yeah. We don't have any modules with non-default sources.... Perhaps we need to make one?

File: nengo/spa/thalamus.py (hunk 2 above)
Comment (xchoo) [inline nengo/spa/thalamus.py]: Ah. Righto. But this does mean that if we are using custom modules, this code won't be able to check for them, yes?

File: nengo/spa/spa.py (hunk 1 above)
Comment (tcstewar) [inline nengo/spa/spa.py]: And, just to clarify my ambiguous comment above:

> So this means that we can't have inputs and outputs names with an underscore in them, yes? 

Yes, but that's always been the case.  That's not something new to this PR.

Comment (xchoo) [inline nengo/spa/spa.py]: Whoopa. See comments in the next block of code (test_thalamus.py). (Accidentally replied in the wrong area)

File: nengo/spa/thalamus.py (hunk 2 above)
Comment (tcstewar) [inline nengo/spa/thalamus.py]: What do you want it to do in those cases?  What it does right now is use the Buffer's geometry for storing information if it can find it, and otherwise it falls back on the parameter its been told to use.

I could see just getting rid of that check and always using the subdim_channel parameter.....  but that'd be a performance change from this PR, and my goal here was just to fix a bug that's breaking Spaun.  :)

**Comment by tbekolay** [comment]
This fix is needed for Spaun, so I gave it a look over. Seems OK? I added a changelog entry (see the FIXUP commit) and fixed up the history a bit; @tcstewar, make sure the changelog entry looks OK. If so, then I'll merge this.

Also, let this comment be a reminder to any who care about this PR, if you want to give it a review, now's the time ;) There was some other discussion about SPA stuff, so if you want to keep that going, make a new issue for it.
